src/main/java/org/truffleruby/core/range/RubyObjectRange.java

```diff
@@ -29,12 +29,20 @@ public RubyObjectRange(RubyClass rubyClass, Shape shape, boolean excludedEnd, Ob
         this.end = end;
     }
 
+    public boolean isNilNil() {
+        return begin == Nil.INSTANCE && end == Nil.INSTANCE;
```

Comment:
Those should be `Primitive.nil?`

Reply:
I haven't figured out how to use Primitives in Java code; could you point me to some part of the codebase where this happens or rewrite this function for me? Thanks!

Reply:
Sorry, this is just me being absent-minded and thinking this was Ruby. It's actually perfectly fine as-is!

Primitives are just a means to avoid method call overhead in Ruby code. They're just Truffle nodes with the `@Primitive` annotation. They can be reused like other nodes in Java (e.g. via `@Cached MyPrimitiveNode` in the parameter list of a method marked `@Specialization`) - but this is just useful to reuse logic, there is no performance gain from using them in Java code.

You can find the code for a given primitive in the generated stubs (e.g. for `Primitive.nil?`, look in `src/main/java/org/truffleruby/core/support/TypeNodes.java`, there's an auto-generated comment pointing at `org.truffleruby.core.support.TypeNodes.IsNilNode`).

For an example of primitive being reused, see for instance `ArrayIndexNodes.ReadNormalizedNode`.

Reply:
Oh awesome example, thanks!!!